### ISSUE TITLE
False positive for useless_conversion when using const generics.

### ISSUE THREAD

**Issue by chanks** [rust-lang/rust-clippy#7205]
Lint name: useless_conversion

[Reproduction using current nightly on Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=680cbd3239749c01f74ec8159d177626)

I tried this code:

```rust
struct MyStruct<const C: char>;

impl From<MyStruct<'a'>> for MyStruct<'b'> {
    fn from(_s: MyStruct<'a'>) -> Self {
        MyStruct
    }
}

fn main() {
    let s: MyStruct<'a'> = MyStruct;
    let _: MyStruct<'b'> = s.into();
}
```
I expected to see this happen: No Clippy warnings.

Instead, this happened:
```
    Checking playground v0.0.1 (/playground)
warning: useless conversion to the same type: `MyStruct<'a'>`
  --> src/main.rs:11:28
   |
11 |     let _: MyStruct<'b'> = s.into();
   |                            ^^^^^^^^ help: consider removing `.into()`: `s`
   |
   = note: `#[warn(clippy::useless_conversion)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.54s
```

The conversion isn't useless - removing `.into()` is a compilation error.

### Meta

Clippy and rustc are current nightly: 2021-05-09

This seems like a very similar error to #7206, but I filed them separately because they're different lints.
